Quick Search Icons

Raemon opened the pull request:
Currently there is no way to search for sequences. A major reason for this is that our current search tool didn't have space to include them.

Current one:
<img width="1728" alt="image" src="https://user-images.githubusercontent.com/3246710/168743625-9c02ffc8-0cd1-4d7a-8f39-7d0d89145d13.png">

This PR removes the large title-headers for each search section, and instead uses icons to distinguish between categories of search results, making space for a 5th content type:

<img width="1728" alt="image" src="https://user-images.githubusercontent.com/3246710/168746251-c51630b0-7efc-4d70-9dfb-92cb4dca336e.png">

(Note: I alternate the icons between outlined and not-outlined to make it easy to scan the sections)

This PR _doesn't_ attempt to improve our full-page-search, which will definitely require a significantly different design to accommodate 5 content types. I'm not yet sure what to do about that, and think it may require getting more into the weeds of how search actually works. 

Looking at other websites with multiple content types, I don't see any sites with multiple columns of content – instead they typically fit all content types into a single column, with a secondary column for auxilary information. I have a few ideas for how to go about this but they'll require a fair amount of thought and work.

I suspect that when we invest more into improving the full-page-search, we'll come up with some improvements that also streamline the results here (possibly figuring out how to discard results that seem to have zero relevance)

darkruby501 commented:
Creative! Totally gets Sequences into the results. I find this kind of unaesthetic though and find the results much harder to read or something, maybe because the changing formatting and it seeming super dense. I'm not sure on net the change is worth it to get sequences into the results, though I'm interested in other people's opinions

jimrandomh commented on packages/lesswrong/components/search/CommentsSearchHit.tsx:
```diff
@@ -3,14 +3,24 @@ import { Link } from '../../lib/reactRouterWrapper';
+    paddingBottom: 8,
+    display: 'flex',
+    alignItems: 'center',
+    borderTop: "solid 1px rgba(0,0,0,.1)"
```
```suggestion
    borderTop: theme.palette.border.faint
```
(a unit tests complains about this because it doesn't handle dark mode)

s-cheng commented on packages/lesswrong/components/search/SearchBarResults.tsx:
```diff
@@ -79,60 +76,43 @@ const SearchBarResults = ({closeSearch, currentQuery, classes}: {
             <Index indexName={getAlgoliaIndexName("Users")}>
-              <div className={classes.header}>
-                <Typography variant="body1">Users</Typography>
-                <Link to={`/search?terms=${currentQuery}`} className={classes.seeAll}>
-                  See all results
-                </Link>
```
Looks like this link is gone now, but it's still useful to have since this dropdown limits the number of results. Maybe include it at the bottom of the list?

s-cheng commented on packages/lesswrong/components/search/TagsSearchHit.tsx:
```diff
@@ -29,8 +39,12 @@ const TagsSearchHit = ({hit, clickAction, classes}: {
+  const { LWTooltip } = Components
   const tag = hit as AlgoliaTag;
   return <div className={classes.root}>
+    <LWTooltip title="Tag">
```
The EA Forum calls these "Topics" now, so you'll need to use the new config that JP created:

`taggingNameIsSet.get() ? taggingNameCapitalSetting.get() : 'Tag'`

jpaddison3 replied:
You can just use taggingNameCapitalSetting.get() as it defaults to Tag

s-cheng replied:
Oh right, I forgot

s-cheng commented on packages/lesswrong/components/search/TagsSearchHit.tsx:
```diff
@@ -1,19 +1,29 @@
+    paddingBottom: 8,
+    display: 'flex',
+    alignItems: 'center',
+    borderTop: "solid 1px rgba(0,0,0,.1)"
```
If there are no "user" search results, then this `border-top` will show up at the top of the dropdown:
<img width="300" alt="Screen Shot 2022-05-17 at 5 27 08 PM" src="https://user-images.githubusercontent.com/9057804/168914052-5c0e4f83-1d32-432d-aa7d-b8e86fca24e1.png">

Perhaps using `border-bottom` would work better, especially if you add the "see all results" link under these results.

Raemon replied:
That ends up causing the same problem – the issue is that I'm handling the UsersSearchHit padding different from everything else, because user search hits are comparatively teeny, and giving them an extra 20px of padding adds up to a lot of wasted space.

So instead, the overall list of users gets a small bit of padding, but then that ends up looking awkward when it's empty. I'll see if I can figure out a solution.

s-cheng replied:
Hmm I see what you mean. I think this is fine for now, but maybe in the future it would be worth refactoring a bit to use [StateResults](https://www.algolia.com/doc/api-reference/widgets/state-results/react/). Then we could hide that extra border when there are no user results.

Raemon replied:
Nod. I thought of some other solutions overnight but think I'd prefer to leave them for a future PR.

s-cheng commented on packages/lesswrong/components/search/SequencesSearchHit.tsx:
```diff
@@ -22,28 +39,38 @@ const styles = (theme: ThemeType): JssStyles => ({
       marginRight: 5,
     }
   },
+  snippet: {
+
+  }
 });
 
 const SequencesSearchHit = ({hit, clickAction, classes}: {
-  hit: Hit<AlgoliaSequence>,
+  hit: Hit<any>,
```
Is `hit` supposed to have the type `AlgoliaSequence`?

Raemon replied:
The other search-hit components don't have types atm. When I try adding the types, it throws errors, I think because the interfaces don't have snippets or highlights. For now I was just copying the typing of the other search-hit components but I could dig into setting it up more properly if that seems important.

Raemon replied:
(It looks like each search hit component has a line like `const post = (hit as AlgoliaPost);` where it's doing some type checking. I don't currently understand the deal with that, flagging it for reference)

s-cheng replied:
Ah ok, yeah prob not worth spending more time on it right now.

darkruby501 replied:
@Raemon Maybe get Robert to take a look at the types. I believe this is something of an area he's specialized in.

s-cheng reviewed (changes requested):
I like how this looks, and I think it's an improvement. :) Just need to get the tests passing again and re-add the "See all results" link.

Raemon commented:
I've fixed a bunch of issues:
– forum-specific tag name (taggingNameCapitalSetting.get() )
– switched to border.faint
– added See All Results to the bottom
– added a little more padding which hopefully will make it feel less dense

I haven't changed the types for the Hit props, nor find a good way for the user results to avoid looking a bit weird when they're empty.

Some additional changes, possibly controversial:
– I made it so tags, posts and sequences only show their snippet if they actually got a search result in the main body content (to save on vertical space)
– I removed the bio field from search for users, since it mostly gave confusing results. I think this'll make it easier to find users with lower karma.
– I reduced the word count for the posts' snippet. (I'm not 100% sure how this will feel, I'm not sure how easy it is to re-run algolia configutation sync, just updated the file)

Raemon commented:
I added optional icons (left them disabled by default, so we don't have to worry about any other places where they might show up awkwardly)

Also fixed a bug where links in the snippet would sometimes appear super-long.

@s-cheng it seems like in your version of User searching you probably actually want to include the bio in a snippet? (which might warrant a different design for the UsersSearchHit, making it more similar to the other search-hits instead of deliberately compressed)

Raemon commented:
Ban User test is failing, not sure if that's expected or not.
